File: scripts/map_media.mjs
#!/usr/bin/env node
/**
 * Extract per-language video and audio manifests from each SE deployment's
 * large "contents" JS chunk, and write them into data/pkf/<iso>/info.json
 * under `media`. Nothing is stored locally — all references are absolute URLs
 * to scriptureearth.org / youtube.com / 4.dbt.io, rendered on demand.
 *
 * Structure stored:
 *   media.audio = {
 *     base_url: "https://www.scriptureearth.org/data/<iso>/audio",
 *     items: [ { bookCode, chapter, filename, url, len, size, timingFile } ]
 *   }
 *   media.videos = [
 *     { id, title, width, height, thumbnail, thumbnailUrl, onlineUrl,
 *       kind: 'youtube' | 'hls' | 'other',
 *       placement: { bookCode, chapter, verse, pos } }
 *   ]
 *
 * Idempotent. Safe to re-run after SE redeploys (JS-chunk hashes change, so
 * we always re-discover the manifest chunk via grep).
 */
import { readFileSync, writeFileSync, readdirSync, statSync } from 'node:fs';
import { join } from 'node:path';

const SE = 'https://scriptureearth.org';
const PKF_ROOT = 'data/pkf';
const IMG_EXT_RE = /\.(jpg|jpeg|png|webp)$/i;
const SW_IMG_RE = /\/_app\/immutable\/assets\/([A-Za-z0-9_\-]+)\.([A-Za-z0-9_\-]+)\.(jpg|jpeg|png|webp)/gi;

// --- low-level utilities ----------------------------------------------------

async function fetchText(url) {
    const r = await fetch(url, { headers: { 'User-Agent': 'bw-map-media/1.0' } });
    if (!r.ok) return null;
    return await r.text();
}

async function headSize(url) {
    try {
        const r = await fetch(url, {
            method: 'HEAD',
            headers: { 'User-Agent': 'bw-map-media/1.0' }
        });
        const cl = r.headers.get('content-length');
        return cl ? parseInt(cl, 10) : 0;
    } catch {
        return 0;
    }
}

/**
 * Walk string from startIdx (which must be '{') and return the index of the
 * matching '}', respecting JS strings so braces inside strings aren't counted.
 */
function findMatchingBrace(text, startIdx) {
    let depth = 0;
    let inStr = false;
    let stringChar = null;
    for (let i = startIdx; i < text.length; i++) {
        const c = text[i];
        if (inStr) {
            if (c === '\\') {
                i++;
                continue;
            }
            if (c === stringChar) inStr = false;
            continue;
        }
        if (c === '"' || c === "'" || c === '`') {
            inStr = true;
            stringChar = c;
            continue;
        }
        if (c === '{') depth++;
        else if (c === '}') {
            depth--;
            if (depth === 0) return i;
        }
    }
    return -1;
}

/** For the given minified JS text and a regex whose first match inside the
 * enclosing object is taken as a handle, return each enclosing {...} object
 * substring (including the outer braces). */
function collectObjectsByHandle(text, handleRe) {
    const out = [];
    handleRe.lastIndex = 0;
    let m;
    while ((m = handleRe.exec(text)) !== null) {
        // Walk backward to find the '{' that starts the enclosing object.
        let idx = m.index;
        let depth = 0;
        while (idx > 0) {
            const c = text[idx];
            if (c === '}') depth++;
            else if (c === '{') {
                if (depth === 0) break;
                depth--;
            }
            idx--;
        }
        if (text[idx] !== '{') continue;
        const end = findMatchingBrace(text, idx);
        if (end > idx) out.push(text.slice(idx, end + 1));
    }
    return out;
}

function getField(objText, key) {
    // Minified JS: keys are usually bare (id:) but sometimes quoted ("id":).
    // Value: either a JS string "…", a number, or a nested object/array — we
    // only extract strings and numbers here.
    const re = new RegExp(
        `(?:^|[\\{,])\\s*["']?${key}["']?\\s*:\\s*("(?:[^"\\\\]|\\\\.)*"|'(?:[^'\\\\]|\\\\.)*'|-?\\d+(?:\\.\\d+)?)`
    );
    const m = objText.match(re);
    if (!m) return null;
    const raw = m[1];
    if (raw.startsWith('"') || raw.startsWith("'")) {
        return JSON.parse(raw.replace(/\\'/g, "'"));
    }
    return parseFloat(raw);
}

/** Find a nested object value for a given key (e.g. "placement"). Returns the
 * substring (including braces) or null. */
function getNestedObject(objText, key) {
    const re = new RegExp(`(?:^|[\\{,])\\s*["']?${key}["']?\\s*:\\s*\\{`);
    const m = objText.match(re);
    if (!m) return null;
    const openIdx = m.index + m[0].length - 1;
    const end = findMatchingBrace(objText, openIdx);
    if (end < 0) return null;
    return objText.slice(openIdx, end + 1);
}

// --- video extraction -------------------------------------------------------

function detectVideoKind(url) {
    if (!url) return 'other';
    if (url.includes('youtube.com') || url.includes('youtu.be')) return 'youtube';
    if (url.includes('.m3u8')) return 'hls';
    if (url.includes('arclight.org')) return 'arclight';
    if (url.includes('vimeo.com')) return 'vimeo';
    if (/\.(mp4|webm|ogv|m4v|mov)(\?|$)/i.test(url)) return 'file';
    return 'other';
}

/** SAB stores some URLs pre-HTML-encoded (e.g. `&amp;` instead of `&`).
 *  Decode the handful of entities we actually see so the URL works verbatim
 *  as an iframe src or a fetch target. */
function decodeHtmlEntities(s) {
    if (!s) return s;
    return s
        .replace(/&amp;/g, '&')
        .replace(/&lt;/g, '<')
        .replace(/&gt;/g, '>')
        .replace(/&quot;/g, '"')
        .replace(/&#39;/g, "'");
}

function parsePlacementRef(refStr) {
    if (!refStr) return {};
    // Accepts "MRK 1:1", "MRK.1.1", "JHN.1.2", etc.
    const m = refStr.match(/^([A-Z1-3]{2,4})[\s.:](\d+)(?:[.:](\d+))?/);
    if (!m) return { rawRef: refStr };
    return {
        bookCode: m[1],
        chapter: parseInt(m[2], 10),
        verse: m[3] != null ? parseInt(m[3], 10) : null
    };
}

function extractVideos(chunkText, imageUrlMap) {
    const videoObjs = collectObjectsByHandle(chunkText, /\bonlineUrl\s*:/g);
    const out = [];
    const seen = new Set();
    for (const obj of videoObjs) {
        const id = getField(obj, 'id');
        const onlineUrl = getField(obj, 'onlineUrl');
        if (!id || !onlineUrl) continue;
        const key = `${id}|${onlineUrl}`;
        if (seen.has(key)) continue;
        seen.add(key);

        const thumbnail = getField(obj, 'thumbnail');
        const placementText = getNestedObject(obj, 'placement');
        const placement = placementText
            ? {
                  ...parsePlacementRef(getField(placementText, 'ref')),
                  pos: getField(placementText, 'pos') ?? null,
                  collection: getField(placementText, 'collection') ?? null
              }
            : {};

        const decodedUrl = decodeHtmlEntities(onlineUrl);
        out.push({
            id,
            title: getField(obj, 'title') ?? '',
            width: getField(obj, 'width') ?? null,
            height: getField(obj, 'height') ?? null,
            thumbnail: thumbnail ?? null,
            thumbnailUrl: thumbnail ? imageUrlMap[thumbnail] ?? null : null,
            onlineUrl: decodedUrl,
            kind: detectVideoKind(decodedUrl),
            placement
        });
    }
    return out;
}

// --- audio extraction -------------------------------------------------------

function extractAudioSources(chunkText) {
    // SE's SAB apps expose these under `sources:{d1:{type:"download",...,address:"http://..."}}`.
    // We anchor on the inner shape rather than the parent key name (which has
    // been observed as `sources`, sometimes `audioSources`) for robustness.
    const re = /\b(?:audio)?[Ss]ources\s*:\s*\{\s*[A-Za-z][A-Za-z0-9_]*\s*:\s*\{\s*type\s*:\s*"(?:download|streaming)"/g;
    const m = re.exec(chunkText);
    if (!m) return {};
    const openIdx = chunkText.indexOf('{', m.index);
    const end = findMatchingBrace(chunkText, openIdx);
    if (end < 0) return {};
    const inner = chunkText.slice(openIdx, end + 1);

    const sources = {};
    const keyRe = /([A-Za-z][A-Za-z0-9_]*)\s*:\s*\{/g;
    let km;
    while ((km = keyRe.exec(inner)) !== null) {
        const k = km[1];
        const start = keyRe.lastIndex - 1;
        const e = findMatchingBrace(inner, start);
        if (e < 0) continue;
        const body = inner.slice(start, e + 1);
        const address = getField(body, 'address');
        if (address) sources[k] = address.replace(/^http:\/\//, 'https://');
        keyRe.lastIndex = e + 1;
    }
    return sources;
}

/** Parse USFM bookCode+chapter out of the canonical filename pattern
 *  "<iso>-NN-BKC-CC.mp3"  →  { bookCode: "BKC", chapter: CC }. */
function parseAudioFilename(filename) {
    const m = filename.match(/-(\d{2})-([A-Z0-9]{3})-(\d+)\.mp3$/i);
    if (!m) return {};
    return { bookCode: m[2].toUpperCase(), chapter: parseInt(m[3], 10) };
}

function extractAudioItems(chunkText, sources) {
    const objs = collectObjectsByHandle(
        chunkText,
        /\bnum\s*:\s*\d+\s*,\s*filename\s*:\s*"[^"]+\.mp3"/g
    );
    const seen = new Set();
    const out = [];
    for (const obj of objs) {
        const filename = getField(obj, 'filename');
        if (!filename || !filename.endsWith('.mp3')) continue;
        if (seen.has(filename)) continue;
        seen.add(filename);
        const src = getField(obj, 'src') ?? '';
        const baseUrl = sources[src] ?? null;
        const url = baseUrl ? `${baseUrl}/${filename}` : null;
        const { bookCode, chapter } = parseAudioFilename(filename);
        out.push({
            filename,
            url,
            bookCode: bookCode ?? null,
            chapter: chapter ?? null,
            num: getField(obj, 'num') ?? null,
            len: getField(obj, 'len') ?? null,
            size: getField(obj, 'size') ?? null,
            timingFile: getField(obj, 'timingFile') ?? null,
            src
        });
    }
    return out;
}

// --- service-worker helpers -------------------------------------------------

function imageUrlMapFrom(iso, swText) {
    const map = {};
    for (const m of swText.matchAll(SW_IMG_RE)) {
        const [, base, hash, ext] = m;
        map[`${base}.${ext}`] = `${SE}/data/${iso}/sab/${iso}/_app/immutable/assets/${base}.${hash}.${ext}`;
    }
    return map;
}

function chunkUrlsFrom(swText) {
    const out = [];
    const seen = new Set();
    for (const m of swText.matchAll(/s\+"(\/_app\/immutable\/(?:chunks|nodes)\/[A-Za-z0-9_.\-]+\.js)"/g)) {
        if (seen.has(m[1])) continue;
        seen.add(m[1]);
        out.push(m[1]);
    }
    return out;
}

// --- main manifest-finder ---------------------------------------------------

async function findManifestChunk(iso, chunkPaths) {
    // HEAD each chunk to get content-length, then try largest-first. The
    // contents manifest is typically the largest chunk by a wide margin.
    const withSizes = await Promise.all(
        chunkPaths.map(async (p) => ({
            url: `${SE}/data/${iso}/sab/${iso}${p}`,
            size: await headSize(`${SE}/data/${iso}/sab/${iso}${p}`)
        }))
    );
    withSizes.sort((a, b) => b.size - a.size);

    // Try chunks in descending size until we find one with onlineUrl or audioSources.
    // Stop after a max of 12 chunks; the manifest is always near the top.
    for (let i = 0; i < Math.min(12, withSizes.length); i++) {
        const body = await fetchText(withSizes[i].url);
        if (!body) continue;
        if (body.includes('onlineUrl:') || body.includes('audioSources:')) {
            return body;
        }
    }
    return null;
}

async function processIso(iso) {
    const isoDir = join(PKF_ROOT, iso);
    const infoPath = join(isoDir, 'info.json');
    if (!statSync(isoDir, { throwIfNoEntry: false })) return null;

    let info;
    try {
        info = JSON.parse(readFileSync(infoPath, 'utf8'));
    } catch {
        return null;
    }

    const sw = await fetchText(`${SE}/data/${iso}/sab/${iso}/service-worker.js`);
    if (!sw) return { iso, error: 'no SW' };

    const imageMap = imageUrlMapFrom(iso, sw);
    const chunks = chunkUrlsFrom(sw);
    if (chunks.length === 0) return { iso, videos: 0, audio: 0 };

    const manifestText = await findManifestChunk(iso, chunks);
    if (!manifestText) return { iso, videos: 0, audio: 0 };

    const videos = extractVideos(manifestText, imageMap);
    const sources = extractAudioSources(manifestText);
    const audioItems = extractAudioItems(manifestText, sources);

    if (videos.length === 0 && audioItems.length === 0) {
        // Clean stale entries
        if ('media' in info) {
            delete info.media;
            writeFileSync(infoPath, JSON.stringify(info, null, 2));
        }
        return { iso, videos: 0, audio: 0 };
    }

    const base_url = sources.d1 ?? Object.values(sources)[0] ?? null;
    info.media = {
        videos,
        audio: {
            base_url,
            items: audioItems
        }
    };
    writeFileSync(infoPath, JSON.stringify(info, null, 2));

    return {
        iso,
        videos: videos.length,
        video_kinds: videos.reduce((acc, v) => {
            acc[v.kind] = (acc[v.kind] || 0) + 1;
            return acc;
        }, {}),
        audio: audioItems.length,
        base_url
    };
}

async function main() {
    const argv = process.argv.slice(2);
    const filter = argv.length ? new Set(argv) : null;
    const concurrency = parseInt(process.env.CONCURRENCY || '6', 10);
    const isos = readdirSync(PKF_ROOT)
        .filter((n) => !n.startsWith('_') && statSync(join(PKF_ROOT, n)).isDirectory())
        .filter((n) => (filter ? filter.has(n) : true))
        .sort();

    let withVideo = 0;
    let withAudio = 0;
    let done = 0;
    const videoKindTotals = {};

    // Worker-pool pattern: N workers pull from a shared queue.
    const queue = isos.slice();
    async function worker() {
        while (queue.length > 0) {
            const iso = queue.shift();
            if (!iso) break;
            try {
                const r = await processIso(iso);
                done++;
                if (!r) continue;
                if (r.error) {
                    console.log(`  ${iso}: ${r.error}  (${done}/${isos.length})`);
                    continue;
                }
                if (r.videos > 0 || r.audio > 0) {
                    if (r.videos) {
                        withVideo++;
                        for (const [k, v] of Object.entries(r.video_kinds || {})) {
                            videoKindTotals[k] = (videoKindTotals[k] || 0) + v;
                        }
                    }
                    if (r.audio) withAudio++;
                    const kinds = Object.entries(r.video_kinds || {})
                        .map(([k, v]) => `${k}=${v}`)
                        .join(',');
                    console.log(
                        `  ${iso.padEnd(6)} videos=${String(r.videos).padStart(3)} [${kinds}]  audio=${String(r.audio).padStart(3)}  (${done}/${isos.length})`
                    );
                } else {
                    // Quiet for no-media langs, but count.
                }
            } catch (e) {
                console.error(`${iso}: ${e.message}`);
            }
        }
    }
    await Promise.all(Array.from({ length: concurrency }, () => worker()));

    console.log(
        `\nDone. ${withVideo} languages with videos, ${withAudio} with audio. Totals by kind: ${JSON.stringify(videoKindTotals)}.`
    );
}

main().catch((e) => {
    console.error(e);
    process.exit(1);
});

File: scripts/map_media.py
#!/usr/bin/env python3
"""
Extract per-language video and audio manifests from each SE deployment's
large "contents" JS chunk, and write them into data/pkf/<iso>/info.json
under `media`. Nothing is stored locally — all references are absolute URLs
to scriptureearth.org / youtube.com / 4.dbt.io, rendered on demand.

Structure stored:
    media.audio = {
      base_url: "https://www.scriptureearth.org/data/<iso>/audio",
      items: [ { bookCode, chapter, filename, url, len, size, timingFile } ]
    }
    media.videos = [
      { id, title, width, height, thumbnail, thumbnailUrl, onlineUrl,
        kind: 'youtube' | 'hls' | 'other',
        placement: { bookCode, chapter, verse, pos } }
    ]

Idempotent. Safe to re-run after SE redeploys (JS-chunk hashes change, so
we always re-discover the manifest chunk via grep).

Env:
    CONCURRENCY=6   number of parallel workers (default 6)
"""
from __future__ import annotations

import json
import os
import re
import sys
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path
from typing import Optional
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

sys.path.insert(0, str(Path(__file__).resolve().parent))
from _lib import REPO_ROOT  # noqa: E402

SE = "https://scriptureearth.org"
PKF_ROOT = REPO_ROOT / "data" / "pkf"

UA = "bw-map-media/1.0"

SW_IMG_RE = re.compile(
    r"/_app/immutable/assets/([A-Za-z0-9_\-]+)\.([A-Za-z0-9_\-]+)\.(jpg|jpeg|png|webp)",
    re.I,
)
CHUNK_IN_SW_RE = re.compile(
    r's\+"(/_app/immutable/(?:chunks|nodes)/[A-Za-z0-9_.\-]+\.js)"'
)


# --- low-level utilities ----------------------------------------------------


def fetch_text(url: str) -> Optional[str]:
    try:
        req = Request(url, headers={"User-Agent": UA})
        with urlopen(req, timeout=30) as r:
            return r.read().decode("utf-8", errors="replace")
    except (HTTPError, URLError, TimeoutError):
        return None


def head_size(url: str) -> int:
    try:
        req = Request(url, headers={"User-Agent": UA}, method="HEAD")
        with urlopen(req, timeout=15) as r:
            cl = r.headers.get("content-length")
            return int(cl) if cl else 0
    except (HTTPError, URLError, TimeoutError):
        return 0


def find_matching_brace(text: str, start_idx: int) -> int:
    """From text[start_idx] == '{', return index of the matching '}', respecting
    JS strings so braces inside strings aren't counted. Returns -1 if no match."""
    depth = 0
    in_str = False
    string_char: Optional[str] = None
    i = start_idx
    n = len(text)
    while i < n:
        c = text[i]
        if in_str:
            if c == "\\":
                i += 2
                continue
            if c == string_char:
                in_str = False
            i += 1
            continue
        if c in ('"', "'", "`"):
            in_str = True
            string_char = c
            i += 1
            continue
        if c == "{":
            depth += 1
        elif c == "}":
            depth -= 1
            if depth == 0:
                return i
        i += 1
    return -1


def collect_objects_by_handle(text: str, handle_re: re.Pattern) -> list[str]:
    """For the first regex match, walk back to find the enclosing '{' and
    forward to find the matching '}'. Return each enclosing object substring."""
    out: list[str] = []
    for m in handle_re.finditer(text):
        # Walk backward to find the '{' that starts the enclosing object.
        idx = m.start()
        depth = 0
        while idx > 0:
            c = text[idx]
            if c == "}":
                depth += 1
            elif c == "{":
                if depth == 0:
                    break
                depth -= 1
            idx -= 1
        if idx < 0 or text[idx] != "{":
            continue
        end = find_matching_brace(text, idx)
        if end > idx:
            out.append(text[idx : end + 1])
    return out


def _decode_js_string(raw: str) -> str:
    """Decode a minified-JS-emitted string literal. Accepts "..." or '...'.
    Handles \\' (minifier artifact for ' inside double-quoted strings)."""
    if raw.startswith(('"', "'")):
        normalized = raw.replace("\\'", "'")
        try:
            return json.loads(normalized)
        except json.JSONDecodeError:
            # Fallback: try converting single-quoted to double-quoted.
            if normalized.startswith("'") and normalized.endswith("'"):
                inner = normalized[1:-1].replace('"', '\\"')
                return json.loads(f'"{inner}"')
            raise
    return raw


def get_field(obj_text: str, key: str):
    """Return the value of `key` (string or number) or None."""
    # Keys may be bare (id:) or quoted ("id":). Values: string or number only.
    pat = re.compile(
        r"(?:^|[{,])\s*[\"']?" + re.escape(key) + r"[\"']?\s*:\s*"
        r"(\"(?:[^\"\\]|\\.)*\"|'(?:[^'\\]|\\.)*'|-?\d+(?:\.\d+)?)"
    )
    m = pat.search(obj_text)
    if not m:
        return None
    raw = m.group(1)
    if raw.startswith(('"', "'")):
        return _decode_js_string(raw)
    try:
        return float(raw) if "." in raw else int(raw)
    except ValueError:
        return None


def get_nested_object(obj_text: str, key: str) -> Optional[str]:
    """Find a nested object for `key` and return the substring including braces."""
    pat = re.compile(r"(?:^|[{,])\s*[\"']?" + re.escape(key) + r"[\"']?\s*:\s*\{")
    m = pat.search(obj_text)
    if not m:
        return None
    open_idx = m.end() - 1
    end = find_matching_brace(obj_text, open_idx)
    if end < 0:
        return None
    return obj_text[open_idx : end + 1]


# --- video extraction -------------------------------------------------------


def detect_video_kind(url: str) -> str:
    if not url:
        return "other"
    if "youtube.com" in url or "youtu.be" in url:
        return "youtube"
    if ".m3u8" in url:
        return "hls"
    if "arclight.org" in url:
        return "arclight"
    if "vimeo.com" in url:
        return "vimeo"
    if re.search(r"\.(mp4|webm|ogv|m4v|mov)(\?|$)", url, re.I):
        return "file"
    return "other"


def decode_html_entities(s: Optional[str]) -> Optional[str]:
    """Decode the small handful of HTML entities SE actually emits."""
    if not s:
        return s
    return (
        s.replace("&amp;", "&")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", '"')
        .replace("&#39;", "'")
    )


def parse_placement_ref(ref: Optional[str]) -> dict:
    if not ref:
        return {}
    m = re.match(r"^([A-Z1-3]{2,4})[\s.:](\d+)(?:[.:](\d+))?", ref)
    if not m:
        return {"rawRef": ref}
    return {
        "bookCode": m.group(1),
        "chapter": int(m.group(2)),
        "verse": int(m.group(3)) if m.group(3) is not None else None,
    }


ONLINE_URL_RE = re.compile(r"\bonlineUrl\s*:")


def extract_videos(chunk_text: str, image_url_map: dict[str, str]) -> list[dict]:
    objs = collect_objects_by_handle(chunk_text, ONLINE_URL_RE)
    out: list[dict] = []
    seen: set[str] = set()
    for obj in objs:
        video_id = get_field(obj, "id")
        online_url = get_field(obj, "onlineUrl")
        if not video_id or not online_url:
            continue
        key = f"{video_id}|{online_url}"
        if key in seen:
            continue
        seen.add(key)

        thumbnail = get_field(obj, "thumbnail")
        placement_text = get_nested_object(obj, "placement")
        placement: dict = {}
        if placement_text:
            placement = {
                **parse_placement_ref(get_field(placement_text, "ref")),
                "pos": get_field(placement_text, "pos"),
                "collection": get_field(placement_text, "collection"),
            }

        decoded = decode_html_entities(online_url) or ""
        out.append(
            {
                "id": video_id,
                "title": get_field(obj, "title") or "",
                "width": get_field(obj, "width"),
                "height": get_field(obj, "height"),
                "thumbnail": thumbnail,
                "thumbnailUrl": image_url_map.get(thumbnail) if thumbnail else None,
                "onlineUrl": decoded,
                "kind": detect_video_kind(decoded),
                "placement": placement,
            }
        )
    return out


# --- audio extraction -------------------------------------------------------


SOURCES_HEAD_RE = re.compile(
    r"\b(?:audio)?[Ss]ources\s*:\s*\{\s*[A-Za-z][A-Za-z0-9_]*\s*:\s*\{\s*type\s*:\s*\"(?:download|streaming)\""
)


def extract_audio_sources(chunk_text: str) -> dict[str, str]:
    m = SOURCES_HEAD_RE.search(chunk_text)
    if not m:
        return {}
    open_idx = chunk_text.find("{", m.start())
    end = find_matching_brace(chunk_text, open_idx)
    if end < 0:
        return {}
    inner = chunk_text[open_idx : end + 1]

    sources: dict[str, str] = {}
    key_re = re.compile(r"([A-Za-z][A-Za-z0-9_]*)\s*:\s*\{")
    pos = 0
    while True:
        km = key_re.search(inner, pos)
        if not km:
            break
        k = km.group(1)
        start = km.end() - 1
        e = find_matching_brace(inner, start)
        if e < 0:
            break
        body = inner[start : e + 1]
        address = get_field(body, "address")
        if address:
            sources[k] = re.sub(r"^http://", "https://", address)
        pos = e + 1
    return sources


AUDIO_ITEM_HANDLE = re.compile(r'\bnum\s*:\s*\d+\s*,\s*filename\s*:\s*"[^"]+\.mp3"')
AUDIO_FILENAME_RE = re.compile(r"-(\d{2})-([A-Z0-9]{3})-(\d+)\.mp3$", re.I)


def parse_audio_filename(filename: str) -> dict:
    m = AUDIO_FILENAME_RE.search(filename)
    if not m:
        return {}
    return {"bookCode": m.group(2).upper(), "chapter": int(m.group(3))}


def extract_audio_items(chunk_text: str, sources: dict[str, str]) -> list[dict]:
    objs = collect_objects_by_handle(chunk_text, AUDIO_ITEM_HANDLE)
    seen: set[str] = set()
    out: list[dict] = []
    for obj in objs:
        filename = get_field(obj, "filename")
        if not filename or not filename.endswith(".mp3"):
            continue
        if filename in seen:
            continue
        seen.add(filename)
        src = get_field(obj, "src") or ""
        base_url = sources.get(src)
        url = f"{base_url}/{filename}" if base_url else None
        parsed = parse_audio_filename(filename)
        out.append(
            {
                "filename": filename,
                "url": url,
                "bookCode": parsed.get("bookCode"),
                "chapter": parsed.get("chapter"),
                "num": get_field(obj, "num"),
                "len": get_field(obj, "len"),
                "size": get_field(obj, "size"),
                "timingFile": get_field(obj, "timingFile"),
                "src": src,
            }
        )
    return out


# --- service-worker helpers -------------------------------------------------


def image_url_map_from(iso: str, sw_text: str) -> dict[str, str]:
    out: dict[str, str] = {}
    for m in SW_IMG_RE.finditer(sw_text):
        base, h, ext = m.group(1), m.group(2), m.group(3)
        out[f"{base}.{ext}"] = f"{SE}/data/{iso}/sab/{iso}/_app/immutable/assets/{base}.{h}.{ext}"
    return out


def chunk_urls_from(sw_text: str) -> list[str]:
    out: list[str] = []
    seen: set[str] = set()
    for m in CHUNK_IN_SW_RE.finditer(sw_text):
        if m.group(1) in seen:
            continue
        seen.add(m.group(1))
        out.append(m.group(1))
    return out


# --- main manifest-finder ---------------------------------------------------


def find_manifest_chunk(iso: str, chunk_paths: list[str]) -> Optional[str]:
    """HEAD each chunk to get content-length, then try largest-first. The
    contents manifest is typically the largest chunk by a wide margin."""
    with_sizes: list[dict] = []
    for p in chunk_paths:
        url = f"{SE}/data/{iso}/sab/{iso}{p}"
        with_sizes.append({"url": url, "size": head_size(url)})
    with_sizes.sort(key=lambda x: -x["size"])

    for entry in with_sizes[: min(12, len(with_sizes))]:
        body = fetch_text(entry["url"])
        if not body:
            continue
        if "onlineUrl:" in body or "audioSources:" in body:
            return body
    return None


def process_iso(iso: str) -> Optional[dict]:
    iso_dir = PKF_ROOT / iso
    info_path = iso_dir / "info.json"
    if not iso_dir.is_dir():
        return None
    try:
        info = json.loads(info_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None

    sw = fetch_text(f"{SE}/data/{iso}/sab/{iso}/service-worker.js")
    if not sw:
        return {"iso": iso, "error": "no SW"}

    image_map = image_url_map_from(iso, sw)
    chunks = chunk_urls_from(sw)
    if not chunks:
        return {"iso": iso, "videos": 0, "audio": 0}

    manifest_text = find_manifest_chunk(iso, chunks)
    if not manifest_text:
        return {"iso": iso, "videos": 0, "audio": 0}

    videos = extract_videos(manifest_text, image_map)
    sources = extract_audio_sources(manifest_text)
    audio_items = extract_audio_items(manifest_text, sources)

    if not videos and not audio_items:
        if "media" in info:
            del info["media"]
            info_path.write_text(json.dumps(info, indent=2, ensure_ascii=False), encoding="utf-8")
        return {"iso": iso, "videos": 0, "audio": 0}

    base_url = sources.get("d1") or (next(iter(sources.values())) if sources else None)
    info["media"] = {
        "videos": videos,
        "audio": {"base_url": base_url, "items": audio_items},
    }
    info_path.write_text(json.dumps(info, indent=2, ensure_ascii=False), encoding="utf-8")

    kinds: dict[str, int] = {}
    for v in videos:
        k = v.get("kind") or "other"
        kinds[k] = kinds.get(k, 0) + 1

    return {
        "iso": iso,
        "videos": len(videos),
        "video_kinds": kinds,
        "audio": len(audio_items),
        "base_url": base_url,
    }


def main() -> int:
    argv = sys.argv[1:]
    filter_set = set(argv) if argv else None
    concurrency = int(os.environ.get("CONCURRENCY", "6"))

    isos = [
        p.name
        for p in sorted(PKF_ROOT.iterdir())
        if p.is_dir() and not p.name.startswith("_") and (not filter_set or p.name in filter_set)
    ]

    with_video = 0
    with_audio = 0
    done = 0
    video_kind_totals: dict[str, int] = {}

    with ThreadPoolExecutor(max_workers=concurrency) as ex:
        futures = {ex.submit(process_iso, iso): iso for iso in isos}
        for fut in as_completed(futures):
            iso = futures[fut]
            try:
                r = fut.result()
            except Exception as e:
                print(f"{iso}: {e}", file=sys.stderr)
                continue
            done += 1
            if not r:
                continue
            if r.get("error"):
                print(f"  {iso}: {r['error']}  ({done}/{len(isos)})")
                continue
            if r.get("videos", 0) or r.get("audio", 0):
                if r.get("videos"):
                    with_video += 1
                    for k, v in (r.get("video_kinds") or {}).items():
                        video_kind_totals[k] = video_kind_totals.get(k, 0) + v
                if r.get("audio"):
                    with_audio += 1
                kinds_str = ",".join(f"{k}={v}" for k, v in (r.get("video_kinds") or {}).items())
                print(
                    f"  {iso:<6} videos={r['videos']:>3} [{kinds_str}]  "
                    f"audio={r['audio']:>3}  ({done}/{len(isos)})"
                )

    print(
        f"\nDone. {with_video} languages with videos, {with_audio} with audio. "
        f"Totals by kind: {json.dumps(video_kind_totals)}."
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
